File: app/src/main/java/com/elementary/tasks/places/create/CreatePlaceActivity.java
package com.elementary.tasks.places.create;

import android.content.ContentResolver;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Toast;

import com.elementary.tasks.R;
import com.elementary.tasks.core.ThemedActivity;
import com.elementary.tasks.core.data.models.Place;
import com.elementary.tasks.core.fragments.AdvancedMapFragment;
import com.elementary.tasks.core.interfaces.MapCallback;
import com.elementary.tasks.core.interfaces.MapListener;
import com.elementary.tasks.core.utils.BackupTool;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.view_models.places.PlaceViewModel;
import com.elementary.tasks.databinding.ActivityCreatePlaceBinding;
import com.google.android.gms.maps.model.LatLng;

import java.io.IOException;
import java.util.ArrayList;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class CreatePlaceActivity extends ThemedActivity implements MapListener, MapCallback {

    private static final int MENU_ITEM_DELETE = 12;

    private ActivityCreatePlaceBinding binding;
    private PlaceViewModel viewModel;
    private AdvancedMapFragment mGoogleMap;
    @Nullable
    private Place mItem;
    @Nullable
    private LatLng place;
    private String placeTitle;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_create_place);
        setSupportActionBar(binding.toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
        binding.toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp);

        mGoogleMap = AdvancedMapFragment.newInstance(false, false, false, false,
                getPrefs().getMarkerStyle(), getThemeUtil().isDark());
        mGoogleMap.setListener(this);
        mGoogleMap.setCallback(this);

        getSupportFragmentManager().beginTransaction()
                .replace(R.id.fragment_container, mGoogleMap)
                .addToBackStack(null)
                .commit();
        loadPlace();
    }

    private void initViewModel(String id) {
        viewModel = ViewModelProviders.of(this, new PlaceViewModel.Factory(getApplication(), id)).get(PlaceViewModel.class);
        viewModel.place.observe(this, place -> {
            if (place != null) {
                showPlace(place);
            }
        });
        viewModel.result.observe(this, commands -> {
            if (commands != null) {
                switch (commands) {
                    case SAVED:
                    case DELETED:
                        finish();
                        break;
                }
            }
        });
    }

    private void loadPlace() {
        Intent intent = getIntent();
        String id = intent.getStringExtra(Constants.INTENT_ID);
        initViewModel(id);
        if (intent.getData() != null) {
            try {
                Uri name = intent.getData();
                String scheme = name.getScheme();
                if (ContentResolver.SCHEME_CONTENT.equals(scheme)) {
                    ContentResolver cr = getContentResolver();
                    mItem = BackupTool.getInstance().getPlace(cr, name);
                } else {
                    mItem = BackupTool.getInstance().getPlace(name.getPath(), null);
                }
                showPlace(mItem);
            } catch (IOException | IllegalStateException e) {
                e.printStackTrace();
            }
        }
    }

    private void showPlace(Place place) {
        this.mItem = place;
        if (place != null) {
            mGoogleMap.addMarker(new LatLng(place.getLatitude(), place.getLongitude()), place.getName(), true, true, -1);
            binding.placeName.setText(place.getName());
        }
    }

    private void addPlace() {
        if (place != null) {
            String name = binding.placeName.getText().toString().trim();
            if (name.matches("")) {
                name = placeTitle;
            }
            if (name == null || name.matches("")) {
                binding.placeName.setError(getString(R.string.must_be_not_empty));
                return;
            }
            Double latitude = place.latitude;
            Double longitude = place.longitude;
            if (mItem != null) {
                mItem.setName(name);
                mItem.setLatitude(latitude);
                mItem.setLongitude(longitude);
            } else {
                mItem = new Place(getPrefs().getRadius(), 0, latitude, longitude, name, "", new ArrayList<>());
            }
            viewModel.savePlace(mItem);
        } else {
            Toast.makeText(this, getString(R.string.you_dont_select_place), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            case R.id.action_add:
                addPlace();
                return true;
            case MENU_ITEM_DELETE:
                deleteItem();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void deleteItem() {
        if (mItem != null) {
            viewModel.deletePlace(mItem);
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (mItem != null && getPrefs().isAutoSaveEnabled()) {
            addPlace();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_palce_edit, menu);
        if (mItem != null) {
            menu.add(Menu.NONE, MENU_ITEM_DELETE, 100, getString(R.string.delete));
        }
        return true;
    }

    @Override
    public void onBackPressed() {
        finish();
    }

    @Override
    public void placeChanged(LatLng place, String address) {
        this.place = place;
        placeTitle = address;
    }

    @Override
    public void onBackClick() {

    }

    @Override
    public void onZoomClick(boolean isFull) {

    }

    @Override
    public void onMapReady() {
        if (mItem != null) showPlace(mItem);
    }
}

File: app/src/main/java/com/elementary/tasks/places/create/CreatePlaceActivity.kt
package com.elementary.tasks.places.create

import android.os.Bundle
import androidx.activity.enableEdgeToEdge
import com.elementary.tasks.R
import com.elementary.tasks.core.data.Commands
import com.elementary.tasks.core.data.ui.place.UiPlaceEdit
import com.elementary.tasks.core.os.PermissionFlowDelegateImpl
import com.github.naz013.ui.common.Dialogues
import com.elementary.tasks.core.utils.ui.trimmedText
import com.elementary.tasks.databinding.ActivityCreatePlaceBinding
import com.elementary.tasks.simplemap.SimpleMapFragment
import com.github.naz013.common.Permissions
import com.github.naz013.common.intent.IntentKeys
import com.github.naz013.domain.Place
import com.github.naz013.ui.common.activity.toast
import com.github.naz013.feature.common.livedata.nonNullObserve
import com.github.naz013.ui.common.activity.BindingActivity
import com.github.naz013.ui.common.view.applyTopInsets
import com.google.android.gms.maps.model.LatLng
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.koin.core.parameter.parametersOf

class CreatePlaceActivity : BindingActivity<ActivityCreatePlaceBinding>() {

  private val dialogues by inject<Dialogues>()
  private val viewModel by viewModel<PlaceViewModel> { parametersOf(getId()) }
  private val permissionFlowDelegate = PermissionFlowDelegateImpl(this)
  private var googleMap: SimpleMapFragment? = null

  override fun inflateBinding() = ActivityCreatePlaceBinding.inflate(layoutInflater)

  override fun onCreate(savedInstanceState: Bundle?) {
    enableEdgeToEdge()
    super.onCreate(savedInstanceState)
    initActionBar()
    googleMap = SimpleMapFragment.newInstance(
      SimpleMapFragment.MapParams(
        isPlaces = false,
        isStyles = true,
        isRadius = true,
        rememberMarkerRadius = false,
        rememberMarkerStyle = false
      )
    )

    googleMap?.mapCallback = object : SimpleMapFragment.MapCallback {
      override fun onMapReady() {
        if (viewModel.canDelete) {
          viewModel.getPlace()?.also {
            showPlaceOnMap(it)
          }
        }
      }

      override fun onLocationSelected(markerState: SimpleMapFragment.MarkerState) {
        viewModel.lat = markerState.latLng.latitude
        viewModel.lng = markerState.latLng.longitude
        viewModel.address = markerState.address
        viewModel.markerStyle = markerState.style
        viewModel.markerRadius = markerState.radius

        if (binding.placeName.trimmedText().isEmpty()) {
          binding.placeName.setText(viewModel.address)
        }
      }
    }

    supportFragmentManager.beginTransaction()
      .replace(R.id.fragment_container, googleMap!!)
      .addToBackStack(null)
      .commit()

    loadPlace()
  }

  private fun initActionBar() {
    binding.appBar.applyTopInsets()
    binding.toolbar.setNavigationOnClickListener { finish() }
    binding.toolbar.setOnMenuItemClickListener { menuItem ->
      when (menuItem.itemId) {
        R.id.action_add -> {
          askCopySaving()
          true
        }

        R.id.action_delete -> {
          dialogues.askConfirmation(this, getString(R.string.delete)) {
            if (it) {
              viewModel.deletePlace()
            }
          }
          true
        }

        else -> false
      }
    }
    updateMenu()
  }

  private fun updateMenu() {
    binding.toolbar.menu.also {
      it.getItem(1).isVisible = viewModel.canDelete
    }
  }

  private fun initViewModel() {
    lifecycle.addObserver(viewModel)
    viewModel.place.nonNullObserve(this) { showPlace(it) }
    viewModel.result.nonNullObserve(this) {
      when (it) {
        Commands.SAVED, Commands.DELETED -> finish()
        else -> {
        }
      }
    }
  }

  private fun getId(): String = intentString(IntentKeys.INTENT_ID)

  private fun loadPlace() {
    initViewModel()
    if (intent.data != null) {
      permissionFlowDelegate.with {
        askPermission(Permissions.READ_EXTERNAL) { readUri() }
      }
    } else if (intent.hasExtra(IntentKeys.INTENT_ITEM)) {
      runCatching {
        viewModel.loadFromIntent(intentSerializable(IntentKeys.INTENT_ITEM, Place::class.java))
      }
    }
  }

  private fun readUri() {
    intent.data?.let { viewModel.loadFromUri(it) }
  }

  private fun showPlace(place: UiPlaceEdit) {
    binding.toolbar.title = getString(R.string.edit_place)
    binding.placeName.setText(place.name)
    showPlaceOnMap(place)
    updateMenu()
  }

  private fun savePlace(newId: Boolean = false) {
    val name: String = binding.placeName.trimmedText()
    if (name.isEmpty()) {
      binding.placeLayout.error = getString(R.string.must_be_not_empty)
      binding.placeLayout.isErrorEnabled = true
      return
    }
    viewModel.savePlace(
      PlaceViewModel.SavePlaceData(
        name = name,
        newId = newId
      )
    )
  }

  private fun askCopySaving() {
    if (viewModel.hasLatLng()) {
      if (viewModel.isFromFile && viewModel.hasSameInDb) {
        dialogues.getMaterialDialog(this)
          .setMessage(R.string.same_place_message)
          .setPositiveButton(R.string.keep) { dialogInterface, _ ->
            dialogInterface.dismiss()
            savePlace(true)
          }
          .setNegativeButton(R.string.replace) { dialogInterface, _ ->
            dialogInterface.dismiss()
            savePlace()
          }
          .setNeutralButton(R.string.cancel) { dialogInterface, _ ->
            dialogInterface.dismiss()
          }
          .create()
          .show()
      } else {
        savePlace()
      }
    } else {
      toast(R.string.you_dont_select_place)
    }
  }

  private fun showPlaceOnMap(place: UiPlaceEdit) {
    googleMap?.run {
      addMarker(
        latLng = LatLng(place.lat, place.lng),
        title = place.name,
        markerStyle = place.marker,
        radius = place.radius,
        clear = true,
        animate = true
      )
    }
  }

  override fun requireLogin(): Boolean {
    return true
  }
}
